connectivity/FEATURE_BLE/source/cordio/source/GattServerImpl.cpp

```diff
@@ -1146,8 +1146,27 @@ uint8_t GattServer::atts_read_cb(
             return read_auth_params.authorizationReply & 0xFF;
         }
 
-        pAttr->pValue = read_auth_params.data;
-        *pAttr->pLen = read_auth_params.len;
+        /* if new data provided copy into the attribute value buffer */
+        if (read_auth_params.data) {
+            if (read_auth_params.offset + read_auth_params.len > pAttr->maxLen) {
+                tr_error("Read authorisation callback set length larger than maximum attribute length. Cannot copy data");
+
+                GattReadCallbackParams read_params = {
+                    connId,
+                    handle,
+                    offset,
+                    read_auth_params.len,
+                    read_auth_params.data,
+                    BLE_ERROR_INVALID_PARAM,
+                };
+                getInstance().handleDataReadEvent(&read_params);
+
+                return ATT_ERR_UNLIKELY;
+            }
+
+            memcpy(pAttr->pValue + read_auth_params.offset, read_auth_params.data, read_auth_params.len);
+            *pAttr->pLen = read_auth_params.len;
```

Comment:
I'm not sure that is correct, the offset should be added. We should clarify in the documentation that the len returned is the length of the buffer returned.

Reply:
I think we should ignore offset altogether

Reply:
I added docs and removed the offset - this way it's clearer. You still need to provide more than offset as this is checked earlier and would otherwise fail so now there's an extra check in the if above.

Reply:
Hang on, this is still not right, I need to allow zero length packets